### PR TITLE
Jvenstad/optional submission data

### PR DESCRIPTION
**Pull request by jonmv**
@freva please reivew.

### CODE REVIEW

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java [hunk 1]
```diff
@@ -39,24 +40,22 @@
...
-        if (authorEmail.isPresent() && ! authorEmail.get().matches("[^@]+@[^@]+")) {
+        if (buildNumber.isEmpty() && (   source.isPresent() || authorEmail.isPresent() || compileVersion.isPresent()
+                                      || buildTime.isPresent() || sourceUrl.isPresent() || commit.isPresent()))
+            throw new IllegalArgumentException("Build number must be present if any other attribute is");
+
+        if (buildNumber.isPresent() && buildNumber.getAsLong() <= 0)
+            throw new IllegalArgumentException("Build number must be > 0");
+
+        if (commit.isPresent() && commit.get().length() > 128)
+            throw new IllegalArgumentException("Commit may not be longer than 128 characters");
```
Comment (freva) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: :clap:

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java [hunk 2]
```diff
@@ -39,24 +40,22 @@
...
-            throw new IllegalArgumentException("both buildNumber and source must be set together");
-        }
-        if (compileVersion.isPresent() != buildTime.isPresent()) {
-            throw new IllegalArgumentException("both compileVersion and buildTime must be set together");
-        }
-        if (buildNumber.isPresent() && buildNumber.getAsLong() <= 0) {
-            throw new IllegalArgumentException("buildNumber must be > 0");
-        }
-        if (authorEmail.isPresent() && ! authorEmail.get().matches("[^@]+@[^@]+")) {
+        if (buildNumber.isEmpty() && (   source.isPresent() || authorEmail.isPresent() || compileVersion.isPresent()
```
Comment (freva) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: When can `buildNumber` be empty?

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java [hunk 3]
```diff
@@ -39,24 +40,22 @@
...
+                                      || buildTime.isPresent() || sourceUrl.isPresent() || commit.isPresent()))
+            throw new IllegalArgumentException("Build number must be present if any other attribute is");
+
+        if (buildNumber.isPresent() && buildNumber.getAsLong() <= 0)
+            throw new IllegalArgumentException("Build number must be > 0");
+
+        if (commit.isPresent() && commit.get().length() > 128)
```
Comment (freva) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: Validate URL somewhere?

**Review by freva — APPROVED** [review]
:fire: :100:

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java (hunk 2 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: `ApplicationVersion.empty`.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java (hunk 3 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: Hmm... Yes, for the explicitly passed `sourceUrl`, I guess that makes sense. I don't dare do it for the generated ones :)

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java (hunk 2 above)
Comment (freva) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: But is it only used internally or is it possible to get this as response to some job? (I know that at least in the past deployments to dev did not have build number)

Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: They explicitly use the `empty` one.

Comment (freva) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: :(

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java (hunk 3 above)
Comment (freva) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: `URI.create()` is a start, but unfortunately it allows relative URLs or just the path component

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java (hunk 2 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: Why sad face?

Comment (freva) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java]: Would be nice to have something we can use to consistently refer to builds in console

**Pull request merged**